llarp/link/link_manager.cpp

```diff
@@ -0,0 +1,359 @@
+        if (rc.IsPublicRouter())
+        {
+          connectedRouters.insert(rc.pubkey);
+        }
+      }
+    };
+
+    ForEachPeer(fn);
+
+    return connectedRouters.size();
+  }
+
+  size_t
+  LinkManager::NumberOfConnectedClients() const
+  {
+    std::set< RouterID > connectedClients;
+
+    auto fn = [&connectedClients](const ILinkSession *session, bool){
+      if (session->IsEstablished())
+      {
+        const RouterContact rc(session->GetRemoteRC());
+        if (!rc.IsPublicRouter())
+        {
+          connectedClients.insert(rc.pubkey);
+        }
+      }
+    };
+
+    ForEachPeer(fn);
+
+    return connectedClients.size();
+  }
+
+  bool
+  LinkManager::GetRandomConnectedRouter(RouterContact &router) const
+  {
+    std::unordered_map< RouterID, RouterContact, RouterID::Hash > connectedRouters;
+
+    ForEachPeer( [&connectedRouters]
+                     (const ILinkSession *peer, bool unused)
+                     {
+                       (void)unused;
+                       connectedRouters[peer->GetPubKey()] = peer->GetRemoteRC();
+                     },
+                 false
+               );
+
+    auto sz = connectedRouters.size();
+    if(sz)
+    {
+      auto itr = connectedRouters.begin();
+      if(sz > 1)
+      {
+        std::advance(itr, randint() % sz);
+      }
+
+      router = itr->second;
+
+      return true;
+    }
+
+    return false;
+  }
+
+  void
+  LinkManager::CheckPersistingSessions(llarp_time_t now)
+  {
+    util::Lock l(&_mutex);
+
+    if (stopping) return;
+
+    auto itr = m_PersistingSessions.begin();
+    while(itr != m_PersistingSessions.end())
+    {
+      auto link = GetLinkWithSessionTo(itr->first);
+      if(now < itr->second)
+      {
+        if(link)
+        {
+          LogDebug("keepalive to ", itr->first);
+          link->KeepAliveSessionTo(itr->first);
+        }
+        else
+        {
+          _sessionMaker->CreateSessionTo(itr->first, nullptr);
+        }
+        ++itr;
+      }
+      else
+      {
+        const RouterID r(itr->first);
+        LogInfo("commit to ", r, " expired");
+        itr = m_PersistingSessions.erase(itr);
+      }
+    }
+  }
+
+  util::StatusObject
+  LinkManager::ExtractStatus() const
+  {
+    std::vector< util::StatusObject > ob_links, ib_links;
+    std::transform(inboundLinks.begin(), inboundLinks.end(),
+                   std::back_inserter(ib_links),
+                   [](const auto &link) -> util::StatusObject {
+                     return link->ExtractStatus();
+                   });
+    std::transform(outboundLinks.begin(), outboundLinks.end(),
+                   std::back_inserter(ob_links),
+                   [](const auto &link) -> util::StatusObject {
+                     return link->ExtractStatus();
+                   });
+
+    util::StatusObject obj{{"outbound", ob_links}, {"inbound", ib_links}};
+
+    return obj;
+  }
+
+  void
+  LinkManager::Init(IOutboundSessionMaker *sessionMaker)
```

Comment:
this should ALSO be the constructor.

Reply:
So for objects which need to point to one another you would end up mixing and matching this style and in-constructor init of pointers.  *Ideally* that situation wouldn't occur, but it's non-trivial to avoid.  In order to have consistency among these new classes I opted for this style.

Reply:
i am segfaulting in my local testing because a timer is called with a member being nullptr and i assumed it's happpening before calling init

Reply:
The `Init`s are run during `Router::FromConfig`.  If that's not early enough then the lack of segfault before was just masking something using `Router` before it was fully configured/initialized.

Reply:
After further discussion, leaving it as-is.

Reply:
Init functions like this are evils in API design :( but okay

Reply:
I can definitely agree with arguments against doing it this way.  It's definitely something to revisit in the future.